core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1244,6 +1245,74 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
     */
   def mask: Stream[F, O] = this.handleErrorWith(_ => Stream.empty)
 
+  /**
+    * Like [[Stream.flatMap]] but interrupts the inner stream when new elements arrive in the outer stream.
+    *
+    * Finializers of each inner stream are guaranteed to run before the next inner stream starts.
+    *
+    * When the outer stream stops gracefully, the currently running inner stream will continue to run.
+    *
+    * When an inner stream terminates/interrupts, nothing happens until the next element arrives
+    * in the outer stream(i.e the outer stream holds the stream open during this time or else the
+    * stream terminates)
+    *
+    * When either the inner or outer stream fails, the entire streams fails.
+    *
+    */
+  def switchMap[F2[x] >: F[x], O2](f: O => Stream[F2, O2])(
+      implicit F2: Concurrent[F2]): Stream[F2, O2] =
+    Stream.eval(Ref.of[F2, Option[Deferred[F2, Unit]]](None)).flatMap { haltRef =>
+      Stream.eval(Deferred[F2, Unit]).flatMap { haltOuter =>
+        // consumes the upstream by sending a packet(stream and a ref) to the next queue and
+        // also does error handling/inturruption/switching by terminating the currently running
+        // inner stream
+        def runUpstream(next: Queue[F2, Option[(O, Deferred[F2, Unit])]],
+                        downstream: Queue[F2, Option[Either[Throwable, Chunk[O2]]]]) =
+          this.attempt
+            .interruptWhen(haltOuter.get.attempt)
+            .onFinalize(next.enqueue1(None))
+            .evalMap {
```

Comment:
I feel like `attempt....evalMap { case Left(e)` is best put in an `handleError` with at the end, so it's easier to focus on the `evalMap` logic in the happy path (the way it is now it's not immediately clear that the `Left` case can only happen once and not per element/chunk)

Reply:
Agree. The main reason I didn't use `handleErrorWith` was because it needs a `Throwable => Stream[F2, O2]` and it just so turns out that `downstream.enqueue1(Left(e).some)` is a `F2[_]` so I thought that fits better in an `evalMap`.

Reply:
Speaking of which, will there ever be a `handleErrorWith` that takes `Throwable => F2[O2]`?

Reply:
it'll have to be a specific Stream method cause it can't fit into `MonadError`. Otoh error handling might not be needed at all here (see my comment below)